File: packages/core/src/transport.ts
import type { DispatchConfig } from "./config";
import type { DispatchEvent, TicketPayload, TicketResponse } from "./types";
import { CONTRACT_VERSION } from "./version";
import { delay } from "./util";

export const QUEUE_LIMIT = 100;
export const DEFAULT_FLUSH_TIMEOUT = 2000;

export interface Transport {
  /** Enqueue an error event for asynchronous delivery (off the hot path). */
  sendEvent(event: DispatchEvent): void;
  /** Synchronously POST a ticket and return the parsed response (or null on failure). */
  postTicket(payload: TicketPayload): Promise<TicketResponse | null>;
  /** Resolve once the queue has drained (or the timeout elapses). */
  flush(timeoutMs?: number): Promise<boolean>;
}

export type FetchLike = (
  input: string,
  init: { method: string; headers: Record<string, string>; body: string; keepalive?: boolean },
) => Promise<{ ok: boolean; status: number; json(): Promise<unknown> }>;

// "dispatch-node/1.2.0 (contract/1)" — lets the backend observe the SDK + contract version.
export function sdkHeader(config: DispatchConfig): string {
  return `${config.sdk.name}/${config.sdk.version} (contract/${CONTRACT_VERSION})`;
}

// A bounded, fire-and-forget transport built on the global fetch (present in Node 18+ and all
// browsers). Mirrors the gem's Transport: a bounded queue (drop-on-overflow, never blocks the
// caller) drained sequentially; failures are swallowed so telemetry never breaks the app.
// Runtime packages can subclass/replace this (Node adds flush-on-exit; browser adds beacons).
export class FetchTransport implements Transport {
  private queue: DispatchEvent[] = [];
  private draining: Promise<void> | null = null;
  private readonly fetchImpl: FetchLike | undefined;

  constructor(
    protected readonly config: DispatchConfig,
    fetchImpl?: FetchLike,
  ) {
    this.fetchImpl =
      fetchImpl ?? (globalThis.fetch as unknown as FetchLike | undefined);
  }

  sendEvent(event: DispatchEvent): void {
    if (this.queue.length >= QUEUE_LIMIT) {
      this.warn(`queue full (${QUEUE_LIMIT}), dropping event ${event.event_id}`);
      return;
    }
    this.queue.push(event);
    this.kick();
  }

  async postTicket(payload: TicketPayload): Promise<TicketResponse | null> {
    const res = await this.post(this.config.endpoint, payload);
    if (!res || !res.ok) return null;
    try {
      return (await res.json()) as TicketResponse;
    } catch {
      return null;
    }
  }

  async flush(timeoutMs: number = DEFAULT_FLUSH_TIMEOUT): Promise<boolean> {
    if (this.draining) {
      await Promise.race([this.draining, delay(timeoutMs)]);
    }
    return this.queue.length === 0;
  }

  private kick(): void {
    if (this.draining) return;
    this.draining = this.drain().finally(() => {
      this.draining = null;
    });
  }

  private async drain(): Promise<void> {
    while (this.queue.length > 0) {
      const event = this.queue.shift()!;
      await this.post(this.config.errorEndpoint, event);
    }
  }

  protected async post(
    url: string,
    body: unknown,
  ): Promise<{ ok: boolean; status: number; json(): Promise<unknown> } | null> {
    if (!this.fetchImpl) {
      this.warn("no fetch implementation available");
      return null;
    }
    try {
      return await this.fetchImpl(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${this.config.apiKey}`,
          "X-Dispatch-Sdk": sdkHeader(this.config),
        },
        body: JSON.stringify(body),
        keepalive: true,
      });
    } catch (err) {
      this.warn(`delivery failed: ${String(err)}`);
      return null;
    }
  }

  protected warn(message: string): void {
    if (this.config.debug) console.warn(`[dispatch] ${message}`);
  }
}

File: python/dispatch-sdk/src/dispatch_sdk/transport.py
"""Async delivery off the hot path: a single daemon worker drains a bounded queue and POSTs
each event. Bounded (drop-on-overflow) so a flood can never grow memory without limit. The
HTTP layer is stdlib urllib (no third-party deps); inject `post_func` to test without sockets.
Mirrors the gem's Transport."""

from __future__ import annotations

import json
import queue
import threading
import time
import urllib.error
import urllib.request
from typing import Any, Callable, Dict, Optional, Tuple

from .config import Config
from .version import CONTRACT_VERSION

QUEUE_LIMIT = 100
READ_TIMEOUT = 5

PostResult = Optional[Tuple[int, str]]
PostFunc = Callable[[str, Dict[str, str], bytes], PostResult]


def sdk_header(config: Config) -> str:
    return f"{config.sdk_name}/{config.sdk_version} (contract/{CONTRACT_VERSION})"


def _dump(obj: Any) -> bytes:
    return json.dumps(obj).encode("utf-8")


class Transport:
    def __init__(self, config: Config, post_func: Optional[PostFunc] = None) -> None:
        self.config = config
        self._post: PostFunc = post_func or self._default_post
        self._queue: "queue.Queue[Dict[str, Any]]" = queue.Queue(maxsize=QUEUE_LIMIT)
        self._worker: Optional[threading.Thread] = None
        self._lock = threading.Lock()

    def send_event(self, event: Dict[str, Any]) -> None:
        try:
            self._queue.put_nowait(event)
        except queue.Full:
            self._warn("queue full, dropping event")
            return
        self._ensure_worker()

    def post_ticket(self, payload: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        result = self._post(self.config.endpoint, self._headers(), _dump(payload))
        if not result:
            return None
        status, text = result
        if not 200 <= status < 300:
            return None
        try:
            parsed = json.loads(text)
            return parsed if isinstance(parsed, dict) else None
        except ValueError:
            return None

    def flush(self, timeout: float = 2.0) -> bool:
        deadline = time.time() + timeout
        while not self._queue.empty() and time.time() < deadline:
            time.sleep(0.005)
        return self._queue.empty()

    def _ensure_worker(self) -> None:
        with self._lock:
            if self._worker is not None and self._worker.is_alive():
                return
            worker = threading.Thread(target=self._run, name="dispatch-transport", daemon=True)
            self._worker = worker
            worker.start()

    def _run(self) -> None:
        while True:
            event = self._queue.get()
            try:
                self._post(self.config.error_endpoint, self._headers(), _dump(event))
            except Exception as err:  # never let a delivery error kill the worker
                self._warn(f"delivery failed: {err}")
            finally:
                self._queue.task_done()

    def _headers(self) -> Dict[str, str]:
        return {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {self.config.api_key}",
            "X-Dispatch-Sdk": sdk_header(self.config),
        }

    def _default_post(self, url: str, headers: Dict[str, str], body: bytes) -> PostResult:
        request = urllib.request.Request(url, data=body, headers=headers, method="POST")
        try:
            with urllib.request.urlopen(request, timeout=READ_TIMEOUT) as response:
                return response.status, response.read().decode("utf-8", "replace")
        except urllib.error.HTTPError as err:
            return err.code, err.read().decode("utf-8", "replace")
        except Exception as err:
            self._warn(f"request failed: {err}")
            return None

    def _warn(self, message: str) -> None:
        if self.config.debug:
            print(f"[dispatch] {message}")
